api/action/index.js

```diff
@@ -0,0 +1,120 @@
+const EMOJI = [
+  'confetti',
+  'suit',
+  'bow',
+  'dog',
+  'skull',
+  'yoyo',
+  'cat',
+];
+
+let buttons = document.querySelectorAll('button');
+
+// The action API does not expose a way to read the action's current enabled/disabled state, so we
+// have to track it ourselves.
+// Relevant feature request: https://bugs.chromium.org/p/chromium/issues/detail?id=1189295
+let actionEnabled = true;
+let showToggleState = document.getElementById('show-toggle-state');
+document.getElementById('toggle-state-button').addEventListener('click', (_event) => {
+  if (actionEnabled) {
+    chrome.action.disable();
+  } else {
+    chrome.action.enable();
+  }
+  actionEnabled = !actionEnabled;
+});
+
+document.getElementById('set-popup-button').addEventListener('click', () => {
+  let popup = document.getElementById('popup-options').value;
+  chrome.action.setPopup({ popup });
+});
+
+let getPopupButton = document.getElementById('get-popup-button');
+getPopupButton.addEventListener('click', () => {
+  chrome.action.getPopup({}, (result) => {
+    document.getElementById('current-popup-value').value = result;
+  });
+});
+
+// Badge Text
+
+document.getElementById('set-badge-text-button').addEventListener('click', () => {
+  let text = document.getElementById('badge-text-input').value;
+  chrome.action.setBadgeText({ text });
+});
+
+document
+  .getElementById('get-badge-text-button')
+  .addEventListener('click', () => {
+    chrome.action.getBadgeText({}, (result) => {
+      document.getElementById('current-badge-text').value = result;
+    });
+  }
+  )
+
+// Badge background color
+
+document.getElementById('set-badge-background-color-button').addEventListener('click', () => {
+  // To show off this method, we must first make sure that there is a badge value
+
+  chrome.action.setBadgeText({ text: 'hi :)' });
+
+  let color = new Array(4);
+  for (let i = 0; i < color.length; i++) {
+    color[i] = Math.floor(Math.random() * 256);
+  }
+
+  chrome.action.setBadgeBackgroundColor({ color });
+});
+
+
+document.getElementById('get-badge-background-color-button').addEventListener('click', () => {
+  chrome.action.getBadgeBackgroundColor({}, (colorArray) => {
+    document.getElementById('current-badge-bg-color').value = JSON.stringify(colorArray, null, 0);
+  });
+});
+
+let lastIconIndex = 0;
+document.getElementById('set-icon-button').addEventListener('click', async () => {
+  lastIconIndex = (lastIconIndex + 1) % EMOJI.length;
+  let emojiFile = `images/emoji-${EMOJI[lastIconIndex]}.png`;
+
+  let response = await fetch(emojiFile);
+  let blob = await response.blob();
+  let imageBitmap = await createImageBitmap(blob);
+  let osc = new OffscreenCanvas(imageBitmap.width, imageBitmap.height);
```

Comment:
This doesn't need to be an OffscreenCanvas—it's on the main thread, right?

Reply:
While true, this specific case was designed to be usable in service workers as well. I'm assuming that developers will copy-paste sample code directly into their extensions, so I wanted to maximize compatibility. Perhaps it would be best to add a comment to that effect?

More generally I've been debating whether or not the demo API interactions should live in the extension's service worker. The main advantage is that this is likely the more general pattern for extension development, but the down side is that this complexity makes the demo harder to understand and that some developers will assume that these APIs can only be used in a background context.

Reply:
I'm happy with the argument that this code might run anywhere and users won't have to worry. 👍